src/libs/Network.js

```diff
@@ -25,25 +26,57 @@ let onError = () => {};
 let isOffline;
 Onyx.connect({
     key: ONYXKEYS.NETWORK,
-    callback: val => isOffline = val && val.isOffline,
-});
-
-let didLoadPersistedRequests;
-Onyx.connect({
-    key: ONYXKEYS.NETWORK_REQUEST_QUEUE,
-    callback: (persistedRequests) => {
-        if (didLoadPersistedRequests || !persistedRequests) {
+    callback: (val) => {
+        if (!val) {
             return;
         }
 
-        // Merge the persisted requests with the requests in memory then clear out the queue as we only need to load
-        // this once when the app initializes
-        networkRequestQueue = [...networkRequestQueue, ...persistedRequests];
-        didLoadPersistedRequests = true;
-        Onyx.set(ONYXKEYS.NETWORK_REQUEST_QUEUE, []);
+        // Client becomes online, process the queue.
+        if (isOffline && !val.isOffline) {
+            // Fake merge which trigger the NETWORK_REQUEST_QUEUE Onyx's subscription callback
+            Onyx.merge(ONYXKEYS.NETWORK_REQUEST_QUEUE, []);
+        }
+        isOffline = val.isOffline;
     },
 });
 
+let didLoadPersistedRequests;
+
+ActiveClientManager.isReady().then(() => {
+    Onyx.connect({
+        key: ONYXKEYS.NETWORK_REQUEST_QUEUE,
+        callback: (persistedRequests) => {
+            // NETWORK_REQUEST_QUEUE is shared across clients, thus every client will have similiar copy of
+            // NETWORK_REQUEST_QUEUE. It is very important to only process the queue from leader client
+            // otherwise requests will be duplicated.
+            // We only process the persisted requests when
+            // a) Client is leader.
+            // b) User is online.
+            // c) requests are not already loaded,
+            // d) When there is at least one request
+            if (!ActiveClientManager.isClientTheLeader()
+                || isOffline
+                || didLoadPersistedRequests
+                || !persistedRequests.length) {
```

Comment:
This line is causing an error for me:

![image](https://user-images.githubusercontent.com/1228807/123339498-ef3a6100-d507-11eb-947e-b9904b2c1067.png)

Reply:
I thought the default value is `[]`, What do you suggest setting a default value on Onyx or check for nullability here.

Reply:
I think adding a `null` check here is OK.